torch/csrc/distributed/rpc/rpc_agent.cpp

```diff
@@ -236,14 +238,20 @@ const WorkerInfo& RpcAgent::getWorkerInfo() const {
   return workerInfo_;
 }
 
-std::shared_ptr<RpcAgent> RpcAgent::currentRpcAgent_ = nullptr;
+std::shared_ptr<RpcAgent>& getRefToCurrentRpcAgent() {
+  #ifdef C10_USE_GLOG
+    google::base::GetLogger(google::GLOG_ERROR);
+  #endif
+  static std::shared_ptr<RpcAgent> currentRpcAgent = nullptr;
```

Comment:
This isn't a leaky singleton, you probably want something like `static RpcAgent* currentRpcAgent` so that it leaks. A full example can be found here: https://isocpp.org/wiki/faq/ctors#static-init-order-on-first-use.

Also, in general it's fragile to try to order initialization of static variables and a refactor could easily break this. We should use a leaky singleton as mentioned above.

Reply:
Yeah, I think here we were in fact going for a different idiom: https://isocpp.org/wiki/faq/ctors#construct-on-first-use-v2. I used the wrong name in the PR summary.

With a true leaky singleton approach the destructor of the agent would _never_ be called, right? I am slightly worried that this may cause problems in some rare cases. For one, the event loop threads would never be joined and would just be destroyed by the OS. I suspect this means that if an error occurs (coming from an external source, like another worker shutting down) during shutdown after the glog mutexes have been destroyed then those threads would still be up and would still attempt to log a message, failing for the same reason. This would only be fully safe if those threads were down by the time the mutexes are destroyed. Also, by never destroying the agent we also never destroy the TensorPipe context, which may mean we may not perform some cleanup operations. At the moment that's not a problem, because all the resources used by TensorPipe are "ephemeral" and will be collected by the kernel, but this may not remain the case forever for future transports.